### PR TITLE
gem-config: add ruby-vips

### PR DESCRIPTION
**Pull request by smaximov**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/ruby-modules/gem-config/default.nix [hunk 1]
```diff
@@ -346,6 +346,22 @@ in
       "--with-ldflags=-L${ncurses.out}/lib"
     ];
   };
+
+  ruby-vips = attrs: {
+    postInstall = ''
+      cd "$(cat $out/nix-support/gem-meta/install-path)"
+
+      substituteInPlace lib/vips.rb \
+        --replace "glib-2.0" "${glib.out}/lib/libglib-2.0${stdenv.hostPlatform.extensions.sharedLibrary}"
```
Comment (alyssais) [inline pkgs/development/ruby-modules/gem-config/default.nix]:
```suggestion
        --replace "glib-2.0" "${glib}/lib/libglib-2.0${stdenv.targetPlatform.extensions.sharedLibrary}"
```

File: pkgs/development/ruby-modules/gem-config/default.nix [hunk 2]
```diff
@@ -346,6 +346,22 @@ in
...
+
+      substituteInPlace lib/vips.rb \
+        --replace "gobject-2.0" "${glib.out}/lib/libgobject-2.0${stdenv.hostPlatform.extensions.sharedLibrary}"
```
Comment (alyssais) [inline pkgs/development/ruby-modules/gem-config/default.nix]:
```suggestion
        --replace "gobject-2.0" "${glib}/lib/libgobject-2.0${stdenv.targetPlatform.extensions.sharedLibrary}"
```

File: pkgs/development/ruby-modules/gem-config/default.nix [hunk 3]
```diff
@@ -346,6 +346,22 @@ in
...
+
+      substituteInPlace lib/vips.rb \
+        --replace "vips_libname = 'vips'" "vips_libname = '${vips.out}/lib/libvips${stdenv.hostPlatform.extensions.sharedLibrary}'"
```
Comment (alyssais) [inline pkgs/development/ruby-modules/gem-config/default.nix]:
```suggestion
        --replace "vips_libname = 'vips'" "vips_libname = '${vips}/lib/libvips${stdenv.targetPlatform.extensions.sharedLibrary}'"
```

**Review by alyssais — CHANGES_REQUESTED** [review]

**Comment by smaximov** [comment]
@alyssais can you explain why you think `hostPlatform` should be replaced with `targetPlatform`? I've read [the manual](https://nixos.org/nixpkgs/manual/#sec-cross-platform-parameters) on platform parameters, and it suggests that `targetPlatform` should only be used for a limited class of packages. The rest of overrides in `defaultGemConfig` uses `hostPlatform`, too.

As for the `"${glib.out}"`&nbsp;&rarr;&nbsp;`"${glib}"` change, it won't work because [`glib`](https://github.com/NixOS/nixpkgs/blob/1056100039733ed9496a9d779207c1c8cb371e1a/pkgs/development/libraries/glib/default.nix#L77) is a [multiple-output package](https://nixos.org/nixpkgs/manual/#chap-multiple-output). `"${glib}"` produces a store path which differs from the store path produced by `"${glib.out}"` and contains GLib binaries only, so using it will break `ruby-vips` (it actually makes my app's test suite fail with `LoadError`).

Since `vips` is a single-output package, `"${vips.out}"` and `"${vips}"` produce the same store path. I've seen both conventions used in the gem config, can I assume that the latter is preferred?

**Comment by alyssais** [comment]
> @alyssais can you explain why you think `hostPlatform` should be replaced with `targetPlatform`? I've read [the manual](https://nixos.org/nixpkgs/manual/#sec-cross-platform-parameters) on platform parameters, and it suggests that `targetPlatform` should only be used for a limited class of packages. The rest of overrides in `defaultGemConfig` uses `hostPlatform`, too.

> hostPlatform
>
>     The "host platform" is the platform on which a package will be run. This is the simplest platform to understand, but also the one with the worst name.

Sorry, I got caught out by the poor naming.

> As for the `"${glib.out}"`&nbsp;&rarr;&nbsp;`"${glib}"` change, it won't work because [`glib`](https://github.com/NixOS/nixpkgs/blob/1056100039733ed9496a9d779207c1c8cb371e1a/pkgs/development/libraries/glib/default.nix#L77) is a [multiple-output package](https://nixos.org/nixpkgs/manual/#chap-multiple-output). `"${glib}"` produces a store path which differs from the store path produced by `"${glib.out}"` and contains GLib binaries only, so using it will break `ruby-vips` (it actually makes my app's test suite fail with `LoadError`).

Also my mistake. Apologies again.

> Since `vips` is a single-output package, `"${vips.out}"` and `"${vips}"` produce the same store path. I've seen both conventions used in the gem config, can I assume that the latter is preferred?

Yeah, the latter is preferred where possible.

Sorry for the errors in my initial review.

**Comment by smaximov** [comment]
> Sorry for the errors in my initial review.

No worries! It made me re-read the relevant parts of the manual and refresh some points that I missed before, so I'd call it a win :wink:

**Comment by smaximov** [comment]
@alyssais, I've replaced `"${vips.out}"` with `"${vips}"`.

**Comment by alyssais** [comment]
Thanks @smaximov!

**Pull request merged**